storage: TestReplicateQueueUpReplicate failed under stress

SHA: https://github.com/cockroachdb/cockroach/commits/35b30e8d95cf0fd111fc98326d92ba4339a9945c

Parameters:
```
TAGS=
GOFLAGS=-race
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stressrace instead of stress and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
make stress TESTS=TestReplicateQueueUpReplicate PKG=github.com/cockroachdb/cockroach/pkg/storage TESTTIMEOUT=5m STRESSFLAGS='-stderr=false -maxtime 20m -timeout 10m'
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=953501&tab=buildLog

```

I181009 09:10:40.395211 113941 base/addr_validation.go:319  [n?] web UI certificate addresses: IP=127.0.0.1,::1; DNS=localhost,*.local; CN=node
I181009 09:10:40.419987 113941 server/config.go:493  [n?] 1 storage engine initialized
I181009 09:10:40.420272 113941 server/config.go:496  [n?] RocksDB cache size: 128 MiB
I181009 09:10:40.420369 113941 server/config.go:496  [n?] store 0: in-memory, size 0 B
I181009 09:10:40.574642 113941 server/node.go:370  [n?] **** cluster 36ac0668-83fe-44a3-9946-0babc481ba37 has been created
I181009 09:10:40.574930 113941 server/server.go:1396  [n?] **** add additional nodes by specifying --join=127.0.0.1:34773
I181009 09:10:40.578193 113941 gossip/gossip.go:393  [n1] NodeDescriptor set to node_id:1 address:<network_field:"tcp" address_field:"127.0.0.1:34773" > attrs:<> locality:<> ServerVersion:<major_val:2 minor_val:0 patch:0 unstable:14 > build_tag:"v2.2.0-alpha.00000000-1458-g35b30e8" started_at:1539076240576290916 
I181009 09:10:40.619316 113941 server/node.go:474  [n1] initialized store [n1,s1]: disk (capacity=512 MiB, available=512 MiB, used=0 B, logicalBytes=7.1 KiB), ranges=1, leases=0, queries=0.00, writes=0.00, bytesPerReplica={p10=7225.00 p25=7225.00 p50=7225.00 p75=7225.00 p90=7225.00 pMax=7225.00}, writesPerReplica={p10=0.00 p25=0.00 p50=0.00 p75=0.00 p90=0.00 pMax=0.00}
I181009 09:10:40.623604 113941 storage/stores.go:242  [n1] read 0 node addresses from persistent storage
I181009 09:10:40.634643 113941 server/node.go:697  [n1] connecting to gossip network to verify cluster ID...
I181009 09:10:40.635256 113941 server/node.go:722  [n1] node connected via gossip and verified as part of cluster "36ac0668-83fe-44a3-9946-0babc481ba37"
I181009 09:10:40.636870 113941 server/node.go:546  [n1] node=1: started with [<no-attributes>=<in-mem>] engine(s) and attributes []
I181009 09:10:40.639948 113941 server/status/recorder.go:609  [n1] available memory from cgroups (8.0 EiB) exceeds system memory 16 GiB, using system memory
I181009 09:10:40.640186 113941 server/server.go:1819  [n1] Could not start heap profiler worker due to: directory to store profiles could not be determined
I181009 09:10:40.650282 113941 server/server.go:1528  [n1] starting https server at 127.0.0.1:44787 (use: 127.0.0.1:44787)
I181009 09:10:40.653836 113941 server/server.go:1530  [n1] starting grpc/postgres server at 127.0.0.1:34773
I181009 09:10:40.653922 113941 server/server.go:1531  [n1] advertising CockroachDB node at 127.0.0.1:34773
W181009 09:10:40.655505 113941 jobs/registry.go:316  [n1] unable to get node liveness: node not in the liveness table
I181009 09:10:40.695742 114072 storage/replica_command.go:298  [n1,split,s1,r1/1:/M{in-ax}] initiating a split of this range at key /System/"" [r2]
I181009 09:10:41.063857 113970 storage/replica_command.go:298  [n1,split,s1,r2/1:/{System/-Max}] initiating a split of this range at key /System/NodeLiveness [r3]
W181009 09:10:41.096692 113916 storage/intent_resolver.go:675  [n1,s1] failed to push during intent resolution: failed to push "unnamed" id=cbb8ee58 key=/Table/SystemConfigSpan/Start rw=true pri=0.01980122 iso=SERIALIZABLE stat=PENDING epo=0 ts=1539076240.779093922,0 orig=1539076240.779093922,0 max=1539076240.779093922,0 wto=false rop=false seq=6
I181009 09:10:41.221725 114128 sql/event_log.go:126  [n1,intExec=optInToDiagnosticsStatReporting] Event: "set_cluster_setting", target: 0, info: {SettingName:diagnostics.reporting.enabled Value:true User:root}
I181009 09:10:41.384792 113920 storage/replica_command.go:298  [n1,split,s1,r3/1:/{System/NodeL…-Max}] initiating a split of this range at key /System/NodeLivenessMax [r4]
I181009 09:10:41.779260 113900 storage/replica_command.go:298  [n1,split,s1,r4/1:/{System/NodeL…-Max}] initiating a split of this range at key /System/tsd [r5]
I181009 09:10:42.006794 114243 storage/replica_command.go:298  [n1,split,s1,r5/1:/{System/tsd-Max}] initiating a split of this range at key /System/"tse" [r6]
I181009 09:10:42.146456 114195 sql/event_log.go:126  [n1,intExec=set-setting] Event: "set_cluster_setting", target: 0, info: {SettingName:version Value:2.0-14 User:root}
```

eek!

```
--- FAIL: TestReplicateQueueUpReplicate (14.84s)
	test_server_shim.go:176: had 21 ranges at startup, expected 20
```

@petermattis, can you work your magic on the log? https://teamcity.cockroachdb.com/repository/download/Cockroach_Nightlies_Stress/953501:id/stress.log

--- a/pkg/config/system.go
+++ b/pkg/config/system.go
@@ -98,6 +98,38 @@ func (s *SystemConfig) Equal(other *SystemConfigEntries) bool {
 	return true
 }
 
+// GetDesc looks for the descriptor value given a key, if a zone is created in
+// a test without creating a Descriptor, a dummy descriptor is returned.
+// If the key is invalid in decoding an ID, GetDesc panics.
+func (s *SystemConfig) GetDesc(key roachpb.Key) *roachpb.Value {
+	if getVal := s.GetValue(key); getVal != nil {
+		return getVal
+	}
+
+	id, err := keys.DecodeDescMetadataID(key)
+	if err != nil {
+		// No ID found for key. No roachpb.Value corresponds to this key.
+		panic(err)
+	}
+
+	testingLock.Lock()
+	_, ok := testingZoneConfig[uint32(id)]
+	testingLock.Unlock()
+
+	if ok {
+		// A test installed a zone config for this ID, but no descriptor.
+		// Synthesize an empty descriptor to force split to occur, or else the
+		// zone config won't apply to any ranges. Most tests that use
+		// TestingSetZoneConfig are too low-level to create tables and zone
+		// configs through proper channels.
+		//
+		// Getting here outside tests is impossible.
+		val := roachpb.MakeValueFromBytes(nil)
+		return &val
+	}
+	return nil
+}
+
 // GetValue searches the kv list for 'key' and returns its
 // roachpb.Value if found.
 func (s *SystemConfig) GetValue(key roachpb.Key) *roachpb.Value {
@@ -410,7 +442,12 @@ func (s *SystemConfig) shouldSplit(ID uint32) bool {
 	shouldSplit, ok := s.mu.shouldSplitCache[ID]
 	s.mu.RUnlock()
 	if !ok {
-		shouldSplit = SplitAtIDHook(ID, s)
+		// Check if the descriptor ID is not one of the reserved
+		// IDs that refer to ranges but not any actual descriptors.
+		shouldSplit = true
+		if ID >= keys.MinUserDescID {
+			shouldSplit = SplitAtIDHook(ID, s)
+		}
 		s.mu.Lock()
 		s.mu.shouldSplitCache[ID] = shouldSplit
 		s.mu.Unlock()

--- a/pkg/config/system_test.go
+++ b/pkg/config/system_test.go
@@ -29,6 +29,8 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
 )
 
+// TODO(benesch): Don't reinvent the key encoding here.
+
 func plainKV(k, v string) roachpb.KeyValue {
 	return kv([]byte(k), []byte(v))
 }
@@ -50,7 +52,7 @@ func sqlKV(tableID uint32, indexID, descriptorID uint64) roachpb.KeyValue {
 }
 
 func descriptor(descriptorID uint64) roachpb.KeyValue {
-	return sqlKV(uint32(keys.DescriptorTableID), 1, descriptorID)
+	return kv(sqlbase.MakeDescMetadataKey(sqlbase.ID(descriptorID)), nil)
 }
 
 func zoneConfig(descriptorID uint32, spans ...config.SubzoneSpan) roachpb.KeyValue {
@@ -358,6 +360,10 @@ func TestComputeSplitKeyTableIDs(t *testing.T) {
 		{subzoneSQL, tkey(start+5, "c"), tkey(start + 6), tkey(start+5, "d")},
 		{subzoneSQL, tkey(start+5, "d"), tkey(start + 6), tkey(start+5, "e")},
 		{subzoneSQL, tkey(start+5, "e"), tkey(start + 6), nil},
+
+		// Testing that no splits are required for IDs that
+		// that do not map to descriptors.
+		{userSQL, tkey(start + 1), tkey(start + 5), nil},
 	}
 
 	cfg := config.NewSystemConfig()

--- a/pkg/sql/sqlbase/system.go
+++ b/pkg/sql/sqlbase/system.go
@@ -33,19 +33,20 @@ func init() {
 // should be considered for a split at all. If it is a database
 // or a view table descriptor, it should not be considered.
 func SplitAtIDHook(id uint32, cfg *config.SystemConfig) bool {
-	descVal := cfg.GetValue(MakeDescMetadataKey(ID(id)))
-	if descVal != nil {
-		var desc Descriptor
-		if err := descVal.GetProto(&desc); err != nil {
+	descVal := cfg.GetDesc(MakeDescMetadataKey(ID(id)))
+	if descVal == nil {
+		return false
+	}
+	var desc Descriptor
+	if err := descVal.GetProto(&desc); err != nil {
+		return false
+	}
+	if dbDesc := desc.GetDatabase(); dbDesc != nil {
+		return false
+	}
+	if tableDesc := desc.GetTable(); tableDesc != nil {
+		if viewStr := tableDesc.GetViewQuery(); viewStr != "" {
 			return false
-		}
-		if dbDesc := desc.GetDatabase(); dbDesc != nil {
-			return false
-		}
-		if tableDesc := desc.GetTable(); tableDesc != nil {
-			if viewStr := tableDesc.GetViewQuery(); viewStr != "" {
-				return false
-			}
 		}
 	}
 	return true

--- a/pkg/storage/client_split_test.go
+++ b/pkg/storage/client_split_test.go
@@ -1227,6 +1227,7 @@ func TestStoreRangeSystemSplits(t *testing.T) {
 	store, _ := createTestStore(t, stopper)
 
 	userTableMax := keys.MinUserDescID + 4
+	var exceptions map[int]struct{}
 	schema := sqlbase.MakeMetadataSchema()
 	// Write table descriptors for the tables in the metadata schema as well as
 	// five dummy user tables. This does two things:
@@ -1278,9 +1279,11 @@ func TestStoreRangeSystemSplits(t *testing.T) {
 			)
 		}
 		for i := keys.MinUserDescID; i <= userTableMax; i++ {
-			expKeys = append(expKeys,
-				testutils.MakeKey(keys.Meta2Prefix, keys.MakeTablePrefix(uint32(i))),
-			)
+			if _, ok := exceptions[i]; !ok {
+				expKeys = append(expKeys,
+					testutils.MakeKey(keys.Meta2Prefix, keys.MakeTablePrefix(uint32(i))),
+				)
+			}
 		}
 		expKeys = append(expKeys, testutils.MakeKey(keys.Meta2Prefix, roachpb.RKeyMax))
 
@@ -1304,12 +1307,13 @@ func TestStoreRangeSystemSplits(t *testing.T) {
 
 	// Write another, disjoint (+3) descriptor for a user table.
 	userTableMax += 3
+	exceptions = map[int]struct{}{userTableMax - 1: {}, userTableMax - 2: {}}
 	if err := store.DB().Txn(context.TODO(), func(ctx context.Context, txn *client.Txn) error {
 		if err := txn.SetSystemConfigTrigger(); err != nil {
 			return err
 		}
-		// This time, only write the last table descriptor. Splits still occur for
-		// every intervening ID. We don't care about the value, just the key.
+		// This time, only write the last table descriptor. Splits only occur for
+		// the descriptor we add. We don't care about the value, just the key.
 		k := sqlbase.MakeDescMetadataKey(sqlbase.ID(userTableMax))
 		return txn.Put(ctx, k, &sqlbase.TableDescriptor{})
 	}); err != nil {

